### PR TITLE
*: sleep 0 when retry whole transaction

### PR DESCRIPTION
**Pull request by lysu**
#### how to reproduce 

use tikv to test, and use default max-txn-time = 590

```
create table t (a int);
insert into t select sleep(600);
```
expect: insert succeed after 600s

but got: forever sleep and got retry over max error.

#### what happens

when commit, tidb finds it's a long transaction(more than 590) and retry transaction but sleep in retry again and again.

#### how it works

when retry transaction, no need to sleep again.

TiDB's auto transaction retry just replay SQL quickly, `sleep(n)` in retry doesn't affect outside world(include business logic or downstream data sync), so we can let them no sleep.

#### what is changed

when retry `preprocess` will let `sleep(1)`, `sleep(1+a)` or `sleep(?)` as `sleep(0)`.

to let `preprocess` method be modified easier, we also do a tiny refactor to its parameter.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - N/A

Related changes

 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/9454)

### CODE REVIEW

File: planner/core/preprocess.go [hunk 1]
```diff
@@ -27,12 +27,24 @@ import (
...
+
 // Preprocess resolves table names of the node, and checks some statements validation.
-func Preprocess(ctx sessionctx.Context, node ast.Node, is infoschema.InfoSchema, inPrepare bool) error {
-	v := preprocessor{is: is, ctx: ctx, inPrepare: inPrepare, tableAliasInJoin: make([]map[string]interface{}, 0)}
+func Preprocess(ctx sessionctx.Context, node ast.Node, is infoschema.InfoSchema, preprocessOpt ...PreprocessOpt) error {
+	var opt PreprocessOpt
+	if len(preprocessOpt) > 0 {
+		opt = preprocessOpt[0]
+	}
+	v := preprocessor{is: is, ctx: ctx, inPrepare: opt.InPrepare, inTxnRetry: opt.InTxRetry, tableAliasInJoin: make([]map[string]interface{}, 0)}
```
Comment (tiancaiamao) [inline planner/core/preprocess.go]: How about this:

```
type PreprocessOpt func(*preprocessor)
func InPrepare(*preprocessor) {
    preprocessor.inPrepare = true
}
func InTxRetry(*preprocessor) {
    preprocessor.inTxnRetry = true
}

... (in func Preprocess)

for _, opt := range preprocessOpt {
    opt(processor)
}
```

File: planner/core/preprocess.go [hunk 2]
```diff
@@ -52,6 +64,7 @@ type preprocessor struct {
 	tableAliasInJoin []map[string]interface{}
 
 	parentIsJoin bool
+	inTxnRetry   bool
```
Comment (tiancaiamao) [inline planner/core/preprocess.go]: inTxnRetry or inTxRetry ... keep the name uniform

File: planner/core/preprocess.go [hunk 3]
```diff
@@ -146,6 +159,13 @@ func (p *preprocessor) Leave(in ast.Node) (out ast.Node, ok bool) {
 			}
 			break
 		}
+
+		// no need sleep when retry transaction and avoid unexpect sleep caused by retry.
+		if p.inPrepare && x.FnName.L == ast.Sleep {
```
Comment (tiancaiamao) [inline planner/core/preprocess.go]: `inTxnRetry` is not used in this commit ?

**Comment by tiancaiamao** [comment]
LGTM

File: planner/core/preprocess.go [hunk 4]
```diff
@@ -27,12 +27,28 @@ import (
...
+// PreprocessOpt presents optional parameters to `Preprocess` method.
+type PreprocessOpt func(*preprocessor)
+
+// InPrepare is a PreprocessOpt that indicates preprocess is executing under prepare statement.
+func InPrepare(p *preprocessor) {
+	p.inPrepare = true
+}
+
+// InTxRetry is a PreprocessOpt that indicates preprocess is executing under transaction retry.
+func InTxRetry(p *preprocessor) {
```
Comment (zz-jason) [inline planner/core/preprocess.go]: s/InTxRetry/InTxnRetry/?

File: planner/core/preprocess.go [hunk 5]
```diff
@@ -52,6 +68,7 @@ type preprocessor struct {
 	tableAliasInJoin []map[string]interface{}
 
 	parentIsJoin bool
+	inTxRetry    bool
```
Comment (zz-jason) [inline planner/core/preprocess.go]: 1. s/inTxRetry/inTxnRetry/?
2. how about putting this variable together with `inPrepare` and adding comments about these two variables?

**Comment by lysu** [comment]
@zz-jason PTAL thx

**Pull request merged**